File: DIP/C#/aplicado/Pedido.cs
public enum FormasPagamento
{
    CARTAOCREDITO
}

public class Pedido
{
    readonly INotificadorService _notificador;
    readonly IProcessadorPagamentoService _processador;
    readonly IReservadorEstoqueService _reservador;
    
    public Pedido(INotificadorService notificador, IProcessadorPagamentoService processador, IReservadorEstoqueService reservador)
    {
        _notificador = notificador;
        _processador = processador;
        _reservador = reservador;
    }
    
    public void Comprar(Carrinho carrinho, DetalhesPagamento detalhesPagamento, bool deveNotificarCliente)
    {
        if(detalhesPagamento.FormaPagamento == FormasPagamento.CARTAOCREDITO)
            _processador.Processar(detalhesDoPagamento, carrinho);
            
        _reservador.ReservarEstoque(carrinho);
        
        if(deveNotificarCliente) 
            _notificador.NotificarCliente(carrinho);
    }
}

File: DIP/Ruby/aplicado/pedido.rb
require 'ostruct'
require 'net/smtp'
require 'notificadorService'
require 'processadorPagamentoService'
require 'reservadorEstoqueService'

class FormasPagamento
    CARTAOCREDITO = 1   
end

class Pedido
    
    def initialize(reservador=ProcessadorPagamentoService.new, notificador=NotificadorService.new, processadorPagamento=ReservadorEstoqueService.new)
        @reservadorEstoque = reservador
        @notificador = notificador
        @processadorPagamento = processadorPagamento
    end

    def comprar(carrinho, detalhesDoPagamento, deveNotificaCliente)
        @processadorPagamento.processar(detalhesDoPagamento, carrinho.valorTotal) if detalhesDoPagamento.formaPagamento == FormasPagamento::CARTAOCREDITO
        
        @reservadorEstoque.reservar(carrinho)

        @notificador.notificarCliente(carrinho) if deveNotificaCliente
    end 
end